Add RFC for deploy control plane without k8s (#4138)
pipe-cd/pipecd: docs/rfcs/0009-control-plane-without-k8s.md

- Start Date: 2022-01-18
- Target Version: 0.41.3

# Summary

Support install PipeCD control plane on other platform which is not k8s

# Motivation

> Now we can deploy the control plane to kubernetes cluster, but some developers that would like to introduce PipeCD can not prepare kubernetes environments. We want to support install PipeCD control plane on platforms other than kubernetes.

Reviewer: ```suggestion
Currently, we can deploy the control plane to Kubernetes cluster, but some developers that would like to introduce PipeCD can not prepare Kubernetes environments. Hence we want to support installing PipeCD control plane on platforms other than Kubernetes.
```

# Detailed design

## 1. Control Plane on docker-compose
    ### Note:

>     - Developers can deploy control plane on a single machine.

Reviewer: ```suggestion
    - Developers can deploy the control plane on a single machine.
```

    - Developers do not have to prepare a datastore and a filestore by themselves or they can easily use a database or a filestore on local machine.
    ### Production Architecture
    - Port-forward from docker-compose to your machine to enable external client access to pipecd-gateway container.
    ![image](assets/control-plane-on-docker-compose.jpg)
    
    
## 2. Control Plane on managed container services (ex. ECS)
    ### Note:

>     - Pipecd give the abstruct architecture for ECS.

Reviewer: ```suggestion
    - PipeCD gives the abstract architecture for ECS.
```

>         - Give a [sample repositpry](https://github.com/pipe-cd/control-plane-aws-ecs-terraform-demo) that describes architecure by using Terraform. 

Reviewer: ```suggestion
        - Give a [sample repository](https://github.com/pipe-cd/control-plane-aws-ecs-terraform-demo) that describes an architecture by using Terraform. 
```

        - The architectures are different by each project, so create your best architecture with refering to this repository and docker-compose file.
    - They can easily use the managed database or storage system on cloud as datastore and filestore.
    
    ### Production Architecture on ECS
    ![image](assets/control-plane-on-aws.jpg)
    
    Note:
    - Devide a pipecd-server and a pipecd-ops to different services because they have the same port and have different authorization.

>     - Pay attention to brocking public access to s3.

Reviewer: ```suggestion
    - Pay attention to brocking public access to S3.
```

        - Add IAM role to ECS to access S3.

>     - Make two type of target group for both of HTTP and GRPC. Use HTTPS for ALB listners because GRPC listner needs HTTPS.

Reviewer: ```suggestion
    - Make two types of target groups for both HTTP and gRPC. Use HTTPS for ALB listeners because gRPC listeners need HTTPS.
```

>     - Set up before containers start to run so that pipecd can use config files and encryption key. You can choose from 3 options as bellow.
>         - Use Secrets Manager as config store.

Reviewer: ```suggestion
    - Set up before containers start to run so that PipeCD can use config files and encryption keys. You can choose from 3 options as below.
        - Use Secrets Manager as config store.
```

        ```

>         # get configuration files from priavete s3 bucket via aws-cli

Reviewer: ```suggestion
        # get configuration files from private S3 bucket via aws-cli
```

        echo $CONTROL_PLANE_CONFIG >> control_plane_config.yaml

>         # replace datastore endpoint (if you use terraform, you can replace by using variable)

Reviewer: ```suggestion
        # replace datastore endpoint (if you use terraform, you can replace by using variables)
```

        sed -i -e s/pipecd-mysql/${var.db_instance_address}/ control-plane-config.yaml

>         # You must set encryption key to environment value

Reviewer: ```suggestion
        # You must set an encryption key to the environment value
```

        echo $ENCRYPTION_KEY >> encryption-key
        # replace cache endpoint (if you use terraform, you can replace by using variable)
        pipecd server \
        --insecure-cookie=true \
        --cache-address=${var.redis_host}:6379 \
        --config-file=control-plane-config.yaml \
        --enable-grpc-reflection=false \
        --encryption-key-file=encryption-key \
        --log-encoding=humanize --metrics=true;
        ```

>         - Use Parameter Store as config store (Attention: Envoy config has over 4096 charactors, so you can not use Parameter Store as eonvy config file store.)
>         - Use S3 as config store. (Attention: install aws-cli to the pipecd-server container to get configuration files from priavete s3 bucket)

Reviewer: ```suggestion
        - Use Parameter Store as config store (Attention: Envoy config has over 4096 characters, so you can not use Parameter Store as envoy config file store.)
        - Use S3 as config store. (Attention: install aws-cli to the pipecd-server container to get configuration files from private s3 bucket)
```

        ```

>         # get configuration files from priavete s3 bucket via aws-cli

Reviewer: ```suggestion
        # get configuration files from private S3 bucket via aws-cli
```

        apk add aws-cli
        aws s3 cp s3://namba-pipecd-control-plane-config/control-plane-config.yaml ./
        ```

# Alternatives

Control Plane without container image

>     - This alternative supports deploy control plane as a binary such as Piped.

Reviewer: ```suggestion
    - This alternative supports deploying the control plane as a binary such as piped.
```

    - It is stressful for developers to set up networking by themselves.

2. Abolish envoy and use ALB instead of envoy for access distribution

>     - We must move configuration for envoy to ALB.

Reviewer: ```suggestion
    - We must move the configuration for envoy to ALB.
```

    - We must make the same number of target groups as pipecd-server has.

> # Unresolved questions
> 
> None

Reviewer: Do we need this section?

```suggestion
```
